src/crypto/algorithms/megolm.ts

```diff
@@ -1419,154 +1438,318 @@ export class MegolmDecryption extends DecryptionAlgorithm {
-                    return;
-                }
+        const roomKey: RoomKey = {
+            senderKey: senderKey,
+            sessionId: content.session_id,
+            sessionKey: content.session_key,
+            extraSessionData,
+            exportFormat: false,
+            roomId: content.room_id,
+            algorithm: content.algorithm,
+            forwardingKeyChain: [],
+            keysClaimed: event.getKeysClaimed(),
+        };
 
-                // If someone sends us an unsolicited key for a room
-                // we're already in, and they're not one of our other
-                // devices or the one who invited us, ignore it
-                if (room && !fromInviter) {
-                    logger.log("forwarded key not from inviter or from us - ignoring");
-                    return;
-                }
-            }
+        return roomKey;
+    }
 
-            exportFormat = true;
-            forwardingKeyChain = Array.isArray(content.forwarding_curve25519_key_chain)
-                ? content.forwarding_curve25519_key_chain
-                : [];
+    /**
+     * Parse a RoomKey out of an `m.forwarded_room_key` event.
+     *
+     * @param event - the event containing the forwarded room key.
+     *
+     * @returns The `RoomKey` if it could be successfully parsed out of the
+     * event.
+     *
+     * @internal
+     *
+     */
+    private forwardedRoomKeyFromEvent(event: MatrixEvent): RoomKey | undefined {
+        const roomKey = this.roomKeyFromEvent(event);
 
-            // copy content before we modify it
-            forwardingKeyChain = forwardingKeyChain.slice();
-            forwardingKeyChain.push(senderKey);
+        if (!roomKey) {
+            return;
+        }
 
-            if (!content.sender_key) {
-                logger.error("forwarded_room_key event is missing sender_key field");
-                return;
-            }
+        const senderKey = event.getSenderKey()!;
+        const content = event.getContent<Partial<IMessage["content"]>>();
 
-            const ed25519Key = content.sender_claimed_ed25519_key;
-            if (!ed25519Key) {
-                logger.error(`forwarded_room_key_event is missing sender_claimed_ed25519_key field`);
-                return;
-            }
+        const senderKeyUser = this.baseApis.crypto!.deviceList.getUserByIdentityKey(olmlib.OLM_ALGORITHM, senderKey);
 
-            keysClaimed = {
-                ed25519: ed25519Key,
-            };
+        // We received this room key from event.getSenderKey(), but the original
+        // creator of the room key is claimed in the content. Let's try to fetch
+        // those keys now.
+        const claimedCurve25519Key = content.sender_key;
+        const claimedEd25519Key = content.sender_claimed_ed25519_key;
 
-            // If this is a key for a room we're not in, don't load it
-            // yet, just park it in case *this sender* invites us to
-            // that room later
```

Comment:
this comment was somewhat useful and has been lost. Reinstate it in `shouldParkForwardedKey` ?

Reply:
Added it back as a doc comment in https://github.com/matrix-org/matrix-js-sdk/pull/2982/commits/fb4d9b662158911dbb5eb5cb8af685a94b55d0fe.